### PR TITLE
Maas 2670 debounce ratio

### PR DESCRIPTION
**Pull request by tetianakh**
Implement debounce ratio (fail if X out of Y last checks failed).

Attempt to refactor the algorithm which determines if a check has failed.

### CODE REVIEW

File: alamo_alerter/alerter.py [hunk 1]
```diff
@@ -66,47 +80,62 @@ def analyze_result(self, results: deque,
         :return: (trigger status, event_type)
         :rtype: tuple
         """
-        last_result = CachedResult(RESULT_UNKNOWN, RESULT_UNKNOWN, False, "")
         result_length = len(results)
         results_list = list(results)
-        if result_length < debounce:
-            # not enough values to proceed with analyze
+        if result_length < debounce and result_length < ratio_out_of:
```
Comment (pnuckowski) [inline alamo_alerter/alerter.py]: Im not sure if we should wait until whole window is ready (` and result_length < ratio_out_of:`)

What if check is failing from the beginning and user set debounce=3 and ratio 4 out of 6.

The trigger will be sent after 6 minutes, but should be sent in 3 minute (debounce is met).

Comment (tetianakh) [reply alamo_alerter/alerter.py]: We are waiting if at least one of the windows is ready. In your example, as soon as we have 3 checks, the `if` condition won't be met and the 3rd check will trigger an alert.

File: alamo_alerter/utils.py [hunk 2]
```diff
@@ -35,6 +44,67 @@ def __init__(self, status, hysteresis, alert_sent, message):
...
+        return all(r.failed for r in self._results[-self._debounce:])
+
+    def previous_result(self):
+        return self._results[-self._debounce - 1].status
+
+    def _pad_with_OK(self, results):
+        return [self._get_OK()] * self._padded_length + results
+
+    @staticmethod
+    def _get_OK():
```
Comment (pnuckowski) [inline alamo_alerter/utils.py]: pep8: function names should be lowercase `_get_ok`

File: alamo_alerter/alerter.py [hunk 3]
```diff
@@ -66,47 +73,62 @@ def analyze_result(self, results: deque,
...
 
-            if latest_result.hysteresis == RESULT_OK:
+            elif latest_result.hysteresis == RESULT_OK:
                 return RESULT_OK, 'resolve'
-        if all([all(result.status in OK_STATUSES for result in results),
-                latest_result.status == RESULT_FAILED]):
-            return RESULT_FAILED, None
-        if all([all_debounce_fail, last_result.status in FAILED_STATUSES]):
+
+        previous_are_OK = all(r.status in OK_STATUSES for r in results)
```
Comment (pnuckowski) [inline alamo_alerter/alerter.py]: pep8: lowercase

File: alamo_alerter/alerter.py [hunk 4]
```diff
@@ -66,47 +73,62 @@ def analyze_result(self, results: deque,
...
+        if all([
+            # debounce conditions:
+            prev_debounce_status in FAILED_STATUSES,  # previously was failing
+            # all checks before last one are failing too
+            all(result.status in FAILED_STATUSES for result in results),
+            latest_result.status == RESULT_OK  # now is OK
+        ]) or all([
+            # ratio conditions:
+            prev_ratio_status in FAILED_STATUSES,  # previously was failing
+            latest_result.status == RESULT_OK,  # now is OK
```
Comment (deejay1) [inline alamo_alerter/alerter.py]: trailing comma

**Pull request merged**